python-package/lightgbm/callback.py

```diff
@@ -363,12 +364,14 @@ def __call__(self, env: CallbackEnv) -> None:
             if self._is_train_set(env.evaluation_result_list[i][0], eval_name_splitted[0], env.model._train_data_name):
                 continue  # train data for lgb.cv or sklearn wrapper (underlying lgb.train)
             elif env.iteration - self.best_iter[i] >= self.stopping_rounds:
-                if self.verbose:
-                    eval_result_str = '\t'.join([_format_eval_result(x, show_stdv=True) for x in self.best_score_list[i]])
-                    _log_info(f"Early stopping, best iteration is:\n[{self.best_iter[i] + 1}]\t{eval_result_str}")
-                    if self.first_metric_only:
-                        _log_info(f"Evaluated only: {eval_name_splitted[-1]}")
-                raise EarlyStopException(self.best_iter[i], self.best_score_list[i])
+                best_score_list = self.best_score_list[i]
+                if best_score_list is not None:
```

Comment:
Why is this new refactoring that you pushed (pulling out this item of `self.best_score_list` and checking if it's `None`) necessary? If the purpose of the PR is to address 4 warnings that are just "need a type annotation"?

Reply:
If we remove the None hint from `_EvalResultTuple` then this part introduces another error.

Reply:
can you please tell me what the new error is that this refactoring helps to avoid?

Changing logic requires more review effort than just-changing-type-hints types of changes. I'm happy to provide that, but I want to understand first what feedback from `mypy` led to this proposed change.

Reply:
These errors:
```
python-package\lightgbm\callback.py:369: error: Item "None" of "Optional[Union[List[Tuple[str, str, float, bool]], List[Tuple[str, str, float, bool, float]]]]" has no attribute "__iter__" (not iterable)  [union-attr]
python-package\lightgbm\callback.py:373: error: Argument 2 to "EarlyStopException" has incompatible type "Optional[Union[List[Tuple[str, str, float, bool]], List[Tuple[str, str, float, bool, float]]]]"; expected "Union[List[Tuple[str, str, float, bool]], List[Tuple[str, str, float, bool, float]]]"  [arg-type]
```

Reply:
hmmm I'm struggling to understand that. 

I don't understand why `mypy` thinks that the type at this point is

```text
Optional[Union[List[Tuple[str, str, float, bool]], List[Tuple[str, str, float, bool, float]]]]
```

I understand how `best_score_list` could *have a `None` as a list element*, but now how it could itself be `None`.

Anyway, I'll try to review this when I can. Now that the logic is changing, it'll take longer to review. Sorry about that.